File: src/PoMode.API/Features/PitchTracking/BasicPitchDecoder.cs
using PoMode.Shared.Analysis;

namespace PoMode.API.Features.PitchTracking;

/// <summary>
/// Converts Basic Pitch's raw onset/frame posterior arrays into discrete <see cref="NoteEvent"/>s.
/// Pure function — no model, no I/O — so it is fully unit-testable with hand-built arrays.
/// </summary>
public static class BasicPitchDecoder
{
    public static IReadOnlyList<NoteEvent> Decode(
        float[,] onsets,
        float[,] frames,
        double framesPerSecond,
        int minMidi,
        double onsetThreshold = 0.5,
        double frameThreshold = 0.3,
        double minDurationSec = 0.058)
    {
        var frameCount = onsets.GetLength(0);
        var pitchCount = onsets.GetLength(1);
        var frameDuration = 1.0 / framesPerSecond;
        var notes = new List<NoteEvent>();

        for (var bin = 0; bin < pitchCount; bin++)
        {
            int? start = null;
            double sumEnergy = 0;
            var count = 0;

            for (var f = 0; f < frameCount; f++)
            {
                var onsetFires = onsets[f, bin] >= onsetThreshold;

                if (start is not null && onsetFires)
                {
                    Emit(notes, bin, minMidi, start.Value, f, sumEnergy, count, frameDuration, minDurationSec);
                    start = null;
                    sumEnergy = 0;
                    count = 0;
                }

                if (start is null && onsetFires)
                {
                    start = f;
                }

                if (start is not null)
                {
                    if (frames[f, bin] >= frameThreshold)
                    {
                        sumEnergy += frames[f, bin];
                        count++;
                    }
                    else
                    {
                        Emit(notes, bin, minMidi, start.Value, f, sumEnergy, count, frameDuration, minDurationSec);
                        start = null;
                        sumEnergy = 0;
                        count = 0;
                    }
                }
            }

            if (start is not null)
            {
                Emit(notes, bin, minMidi, start.Value, frameCount, sumEnergy, count, frameDuration, minDurationSec);
            }
        }

        return [.. notes.OrderBy(n => n.StartSec).ThenBy(n => n.MidiPitch)];
    }

    private static void Emit(
        List<NoteEvent> notes,
        int bin,
        int minMidi,
        int startFrame,
        int endFrame,
        double sumEnergy,
        int count,
        double frameDuration,
        double minDurationSec)
    {
        var durationSec = (endFrame - startFrame) * frameDuration;
        if (durationSec < minDurationSec)
        {
            return;
        }

        var meanEnergy = count > 0 ? sumEnergy / count : 0;
        var velocity = Math.Clamp((int)(meanEnergy * 127), 1, 127);
        var startSec = startFrame * frameDuration;
        notes.Add(new NoteEvent(bin + minMidi, startSec, durationSec, velocity));
    }
}

File: src/PoMode.Client/wwwroot/js/infer/basic-pitch-decoder.js
// The JS side of spec §5's "specified once, implemented twice" contract. This is a direct port of
// PoMode.API/Features/PitchTracking/BasicPitchDecoder.cs — same thresholds, same emit rules, same
// ordering. Both implementations are asserted against the same fixture file
// (tests/TestCommon/basic-pitch-decoder.fixture.json); change one side and the other's test fails
// naming the note. Plain ES module: no npm, no bundler.

/**
 * Converts Basic Pitch's raw onset/frame posterior arrays into discrete note events.
 * @param {number[][]} onsets  [frame][pitchBin] onset posteriors
 * @param {number[][]} frames  [frame][pitchBin] frame (note-activation) posteriors
 * @param {number} framesPerSecond
 * @param {number} minMidi  MIDI pitch of bin 0
 * @returns {{midiPitch: number, startSec: number, durationSec: number, velocity: number}[]}
 */
export function decodeNotes(
    onsets,
    frames,
    framesPerSecond,
    minMidi,
    onsetThreshold = 0.5,
    frameThreshold = 0.3,
    minDurationSec = 0.058) {
    const frameCount = onsets.length;
    const pitchCount = frameCount > 0 ? onsets[0].length : 0;
    const frameDuration = 1.0 / framesPerSecond;
    const notes = [];

    const emit = (bin, startFrame, endFrame, sumEnergy, count) => {
        const durationSec = (endFrame - startFrame) * frameDuration;
        if (durationSec < minDurationSec) {
            return;
        }
        const meanEnergy = count > 0 ? sumEnergy / count : 0;
        // C#'s (int) cast truncates toward zero; Math.trunc is the same operation.
        const velocity = Math.min(127, Math.max(1, Math.trunc(meanEnergy * 127)));
        notes.push({
            midiPitch: bin + minMidi,
            startSec: startFrame * frameDuration,
            durationSec,
            velocity,
        });
    };

    for (let bin = 0; bin < pitchCount; bin++) {
        let start = null;
        let sumEnergy = 0;
        let count = 0;

        for (let f = 0; f < frameCount; f++) {
            const onsetFires = onsets[f][bin] >= onsetThreshold;

            if (start !== null && onsetFires) {
                emit(bin, start, f, sumEnergy, count);
                start = null;
                sumEnergy = 0;
                count = 0;
            }

            if (start === null && onsetFires) {
                start = f;
            }

            if (start !== null) {
                if (frames[f][bin] >= frameThreshold) {
                    sumEnergy += frames[f][bin];
                    count++;
                } else {
                    emit(bin, start, f, sumEnergy, count);
                    start = null;
                    sumEnergy = 0;
                    count = 0;
                }
            }
        }

        if (start !== null) {
            emit(bin, start, frameCount, sumEnergy, count);
        }
    }

    // Same ordering as the C# decoder: by start time, then by pitch. Array.prototype.sort is
    // stable per the ECMAScript spec, matching LINQ's stable OrderBy/ThenBy.
    notes.sort((a, b) => a.startSec - b.startSec || a.midiPitch - b.midiPitch);
    return notes;
}
